NIFI-14426: Add support for HSSF format in ExcelReader processor

zhtk opened the pull request:
<!-- Licensed to the Apache Software Foundation (ASF) under one or more -->
<!-- contributor license agreements.  See the NOTICE file distributed with -->
<!-- this work for additional information regarding copyright ownership. -->
<!-- The ASF licenses this file to You under the Apache License, Version 2.0 -->
<!-- (the "License"); you may not use this file except in compliance with -->
<!-- the License.  You may obtain a copy of the License at -->
<!--     http://www.apache.org/licenses/LICENSE-2.0 -->
<!-- Unless required by applicable law or agreed to in writing, software -->
<!-- distributed under the License is distributed on an "AS IS" BASIS, -->
<!-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. -->
<!-- See the License for the specific language governing permissions and -->
<!-- limitations under the License. -->

# Summary

[NIFI-14426](https://issues.apache.org/jira/browse/NIFI-14426): Add support for HSSF format in ExcelReader processor

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [X] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [X] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [X] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [X] Pull Request based on current revision of the `main` branch
- [ ] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [X] Build completed using `mvn clean install -P contrib-check`
  - [X] JDK 21

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

zhtk commented:
Yes, rebased @exceptionfactory

exceptionfactory commented on nifi-extension-bundles/nifi-poi-bundle/nifi-poi-services/src/main/java/org/apache/nifi/excel/RowIterator.java:
```diff
@@ -42,13 +51,26 @@ class RowIterator implements Iterator<Row>, Closeable {
+                    .setReadSharedFormulas(true) // NOTE: If not set to true, then data with shared formulas fail.
+                    .open(in);
+        } else {
+            // Providing the password to the HSSFWorkbook is done by setting a thread variable managed by
+            // Biff8EncryptionKey. After the workbook is created, the thread variable can be cleared.
+            Biff8EncryptionKey.setCurrentUserPassword(configuration.getPassword());
```
This is problematic for multi-threaded use of ExcelReader, which is very possible. Under the circumstances, it seems best to avoid attempting to support encrypted XLS files right now. This will likely throw exceptions when `HSSFWorkbook` attempts to read, but that is acceptable under the circumstances, as it is a limitation of the library.

exceptionfactory replied:
Although making the Input File Type a configurable property is an option, there are positives and negatives to consider requiring the configuration. On the one hand, IdentifyMimeType could be used with a RouteOnAttribute Processor to determine the appropriate routing. On the other hand, that increases the potential flow design complexity when attempting to support both XLS and XLSX.

I agree that introducing a new configuration property would simplify this logic and remove the need for File Type detection.

Given the mismatch in functionality, with and without support for password-protection, there does seem to be some value in requiring one type or the other.

It is also worth noting that XLS has not been supported previously, and that it has different memory usage characteristics as previously discussed.

With that in mind, I would be open to adding a new property named `Input File Type` with `XLSX` and `XLS` values, where `XLSX` would be the default type. The Property could support Expression Language, which is resolvable from the Map of variables passed to `createRecordReader()`.

dan-s1 replied:
@zhtk Are you clear on the changes needed?

exceptionfactory replied:
@zhtk The thread safety issue is that calling a static method applies to all instances of the Reader. That means if there are multiple instances of the Excel Reader configured, with different passwords, different threads could step on each other. I recommend taking a closer look at the POI library to see if there are any other options. If this is the only approach supported in the library itself, then it would require more significant custom code to implement a solution that works in the context of NiFi.

zhtk replied:
@exceptionfactory, okay, the method is static, and the variable to which the method writes is static, but it's instance of `ThreadLocal<String>`. According to my understanding, any reads/writes to this variable will be only visible by thread accessing the variable.

exceptionfactory replied:
Thanks for pointing to the source implementation of `Biff8EncryptionKey`, I see the use of `ThreadLocal` and the associated comments. That makes a significant difference, and alleviates the concerns of setting a shared variable.

With that clarification, I withdraw my earlier concerns. I would simply add a comment indicating that Biff8EncryptionKey stores the value in a ThreadLocal variable, and proceed.

exceptionfactory commented on nifi-extension-bundles/nifi-poi-bundle/nifi-poi-services/src/main/java/org/apache/nifi/excel/RowIterator.java:
```diff
@@ -129,4 +151,37 @@ private boolean hasExhaustedRows() {
         return exhausted;
     }
+
+    private boolean isXSSFExcelFile(InputStream in, final String password) {
```
Minor naming recommend:
```suggestion
    private boolean isXmlFormat(final InputStream in, final String password) {
```

dan-s1 replied:
or perhaps
```suggestion
    private boolean isXSSF(InputStream in, final String password) {
```

exceptionfactory commented on nifi-extension-bundles/nifi-poi-bundle/nifi-poi-services/src/main/java/org/apache/nifi/excel/RowIterator.java:
```diff
@@ -129,4 +151,37 @@ private boolean hasExhaustedRows() {
+            throw new ProcessException("Failed to determine file type", e);
+        }
+
+        if (fileType == FileMagic.OOXML) {
+            // Unencrypted XLSX file
+            return true;
+        }
+        if (fileType != FileMagic.OLE2) {
+            // Not an Excel file
+            return false;
+        }
+        if (password == null) {
+            // If there is no password on the OLE2 file, then it's an XLS file. Otherwise, the type must be further
+            // investigated.
+            return false;
+        }
```
It would be helpful to refactor this method to use a single return.

exceptionfactory commented on nifi-extension-bundles/nifi-poi-bundle/nifi-poi-services/src/main/java/org/apache/nifi/excel/RowIterator.java:
```diff
@@ -129,4 +151,37 @@ private boolean hasExhaustedRows() {
+        DirectoryNode root;
+        try {
+            root = new POIFSFileSystem(new NonCloseableInputStream(in)).getRoot();
+            in.reset();
```
It looks like the `reset()` call should be moved to a `finally` block in case of exceptions.

dan-s1 replied:
@exceptionfactory Just want to clarify since `reset` can also throw an `IOException` the `finally` would have to also have a try catch block.

exceptionfactory replied:
Yes, that is correct.

exceptionfactory reviewed (changes requested):
Thanks for working on this @zhtk. The general approach looks straightforward. Although there are potential memory implications for the older XLS format, this seems to be the best implementation strategy for now.

I noted one particular issue related to password handling for XLS files, and the lack of thread safety. As mentioned, I think the best option for now is to not support encrypted XLS, and let the read operation fail.

dan-s1 commented on nifi-extension-bundles/nifi-poi-bundle/nifi-poi-services/src/main/java/org/apache/nifi/excel/RowIterator.java:
```diff
@@ -129,4 +151,37 @@ private boolean hasExhaustedRows() {
+            return false;
+        }
+
+        DirectoryNode root;
+        try {
+            root = new POIFSFileSystem(new NonCloseableInputStream(in)).getRoot();
```
In addition I think the creation of `POIFSFileSystem` and the assignment to `root` should be done on separate lines. Also the return statement should be all done in the try block as if the instantiation of `POIFSFileSystem` throws an IOException the return statement is never reached.

```suggestion
        DirectoryNode root;
        try {
           final POIFSFileSystem poifsFileSystem = new POIFSFileSystem(new NonCloseableInputStream(in)).getRoot();
           final DirectoryNode root = poifsFileSystem.getRoot() ;
           return root.hasEntryCaseInsensitive(DEFAULT_POIFS_ENTRY) || root.hasEntryCaseInsensitive(OOXML_PACKAGE);
           }
```

zhtk commented:
After reading whole discussion I decided to remove type detection and to add `Input File Type` property. There are 2 reasons for that:
1. In case of passworded Excel file, to determine file type the whole Flow File stream must be read. At the moment the stream is resettable only up to 1024 * 1024 mark (i.e. in `createRecordReader` there is a call `in.mark(1024 * 1024);`). Reseting stream on files larger than 1 MB could be problematic. Working around this problem would require contributions to both POI and StreamingReader libraries.
2. Performance characteristics of reading XLS files is different than XLSX.

dan-s1 commented:
@zhtk From message above it seems there are conflicts you must resolve

dan-s1 commented on nifi-extension-bundles/nifi-poi-bundle/nifi-poi-services/src/main/java/org/apache/nifi/excel/ExcelReader.java:
```diff
@@ -53,13 +53,13 @@
         + "See Controller Service's Usage for further documentation. "
-        + "This reader is currently only capable of processing .xlsx "
-        + "(XSSF 2007 OOXML file format) Excel documents and not older .xls (HSSF '97(-2007) file format) documents.")
+        + "This reader is capable of processing .xlsx (XSSF 2007 OOXML file format) "
+        + "Excel documents and older .xls (HSSF '97(-2007) file format) documents.")
```
```suggestion
        + "This reader is capable of processing both password and non password protected .xlsx (XSSF 2007 OOXML file format) "
        +  " and older .xls (HSSF '97(-2007) file format) Excel documents.")
```

zhtk replied:
Changed

dan-s1 commented on nifi-extension-bundles/nifi-poi-bundle/nifi-poi-services/src/main/java/org/apache/nifi/excel/InputFileType.java:
```diff
@@ -0,0 +1,47 @@
+import org.apache.nifi.components.DescribedValue;
+
+public enum InputFileType implements DescribedValue {
+    XLS("XLS", "An XLS file"),
+    XLSX("XLSX", "An XLSX file");
```
```suggestion
    XLS("XLS", "HSSF '97(-2007) file format"),
    XLSX("XLSX", ""XSSF 2007 OOXML file format"");
```

exceptionfactory replied:
The `HSSF` acronym should not be used in user-facing labels and documentation. It is an acronym internal to the Apache POI project, and a pejorative one.

exceptionfactory replied:
Yes, including the years in the description is helpful.

zhtk replied:
Good point, changed

dan-s1 commented on nifi-extension-bundles/nifi-poi-bundle/nifi-poi-services/src/main/java/org/apache/nifi/excel/RowIterator.java:
```diff
@@ -42,13 +44,26 @@ class RowIterator implements Iterator<Row>, Closeable {
-                .setAvoidTempFiles(configuration.isAvoidTempFiles())
-                .setReadSharedFormulas(true) // NOTE: If not set to true, then data with shared formulas fail.
-                .open(in);
+        if (configuration.getInputFileType() == InputFileType.XLSX) {
```
This comment is regarding the whole if else statements between lines 47-66. Instead of an if else, an assignment with a switch statement would be clearer to indicate what code is needed for each Excel type. I was thinking along the line of this:

```
this.workbook = switch(configuration.getInputFileType()) {
            case XLSX -> StreamingReader.builder()
                    .rowCacheSize(100)
                    .bufferSize(4096)
                    .password(configuration.getPassword())
                    .setAvoidTempFiles(configuration.isAvoidTempFiles())
                    .setReadSharedFormulas(true) // NOTE: If not set to true, then data with shared formulas fail.
                    .open(in);
            case XLS -> {
                // Providing the password to the HSSFWorkbook is done by setting a thread variable managed by
                // Biff8EncryptionKey. After the workbook is created, the thread variable can be cleared.
                Biff8EncryptionKey.setCurrentUserPassword(configuration.getPassword());
                try {
                    yield new HSSFWorkbook(in);
                } catch (final IOException e) {
                    throw new ProcessException("Failed to open XLS file", e);
                } finally {
                    Biff8EncryptionKey.setCurrentUserPassword(null);
                }
            }
        };
```

dan-s1 replied:
@exceptionfactory Although I believe this is clearer code it would necessitate making changes in TestExcelHeaderSchemaStrategy to account for setting the InputFileType in the context otherwise the unit tests there will fail. Is this still worth it?

exceptionfactory replied:
I'm not quite following the question, is what still worth it? Are you referring to "if-else" versus "switch", adding the new Input File Type property?

dan-s1 replied:
I meant the switch. I feel its clearer seeing which code gets called for which type.

exceptionfactory replied:
I prefer if-else for a smaller numbers of options, and switch when it comes to larger numbers. I would leave it as-is, but would be fine either way.

zhtk replied:
Keeping as is, but switch also makes sense here

dan-s1 commented on nifi-extension-bundles/nifi-poi-bundle/nifi-poi-services/src/test/java/org/apache/nifi/excel/TestExcelRecordReader.java:
```diff
@@ -122,20 +112,32 @@ public void testNonExcelFile() {
-        MalformedRecordException mre = assertThrows(MalformedRecordException.class, () -> new ExcelRecordReader(configuration, getInputStream("olderFormat.xls"), logger));
-        assertTrue(ExceptionUtils.getStackTrace(mre).contains("data appears to be in the OLE2 Format"));
+    public void testOlderExcelFormatFile() throws MalformedRecordException {
+        final List<RecordField> fields = Arrays.asList(
```
```suggestion
        final List<RecordField> fields = List.of(
```

zhtk replied:
Changed

dan-s1 commented on nifi-extension-bundles/nifi-poi-bundle/nifi-poi-services/src/test/java/org/apache/nifi/excel/TestExcelRecordReader.java:
```diff
@@ -318,17 +328,47 @@ void testPasswordProtected() throws Exception {
         assertEquals(DATA.length, records.size());
     }
 
+    @Test
+    void testPasswordProtectedOlderExcelFormat() throws Exception {
+        RecordSchema schema = getPasswordProtectedSchema();
+        ExcelRecordReaderConfiguration configuration = new ExcelRecordReaderConfiguration.Builder()
+                .withSchema(schema)
+                .withPassword(PASSWORD)
+                .withAvoidTempFiles(true)
+                .withInputFileType(InputFileType.XLS)
+                .build();
+
+        InputStream inputStream = new ByteArrayInputStream(PASSWORD_PROTECTED_OLDER_EXCEL.toByteArray());
+        ExcelRecordReader recordReader = new ExcelRecordReader(configuration, inputStream, logger);
+        List<Record> records = getRecords(recordReader, false, false);
+
+        assertEquals(DATA.length, records.size());
+    }
```
Instead of a new test, combine this test and the previous test line 314-329 into a JUnit 5 org.junit.jupiter.params.ParameterizedTest with an org.junit.jupiter.params.provider.EnumSource. You can use the enum value in the configuration and to choose which password protected workbook to use.

zhtk replied:
Changed

dan-s1 commented on nifi-extension-bundles/nifi-poi-bundle/nifi-poi-services/src/test/java/org/apache/nifi/excel/TestExcelRecordReader.java:
```diff
@@ -318,17 +328,47 @@ void testPasswordProtected() throws Exception {
         assertThrows(Exception.class, () -> new ExcelRecordReader(configuration, inputStream, logger));
     }
 
+    @Test
+    void testPasswordProtectedOlderExcelWithoutPassword() {
+        RecordSchema schema = getPasswordProtectedSchema();
+        ExcelRecordReaderConfiguration configuration = new ExcelRecordReaderConfiguration.Builder()
+                .withSchema(schema)
+                .withInputFileType(InputFileType.XLS)
+                .build();
+
+        InputStream inputStream = new ByteArrayInputStream(PASSWORD_PROTECTED_OLDER_EXCEL.toByteArray());
+        assertThrows(Exception.class, () -> new ExcelRecordReader(configuration, inputStream, logger));
+    }
```
Same as previous comment make this a Paramaterized test by combining this test with the one on lines 348-358

dan-s1 reviewed (commented):
@zhtk great work so far. I had a few more suggestions.

dan-s1 commented:
@zhtk Can you please fix the conflicts you have locally and then push?

zhtk commented:
TBH I don't see conflicts, github interface says: "No conflicts with base branch. Changes can be cleanly merged.", and local merge also was performed cleanly. To make sure everything is ok, I have rebased and squashed all commits on the branch, hope now it's okay

dan-s1 commented:
> TBH I don't see conflicts, github interface says: "No conflicts with base branch. Changes can be cleanly merged.", and local merge also was performed cleanly. To make sure everything is ok, I have rebased and squashed all commits on the branch, hope now it's okay

Things look aky now with the force push. Thanks!

dan-s1 reviewed (commented):
I noticed one more thing (hopefully the last :) )

dan-s1 commented:
@exceptionfactory I approve these changes. Is there anything else you see which needs changing?

exceptionfactory reviewed (approved):
Thanks for bringing this to completion @zhtk, and thanks for the reviews along the way @dan-s1, the latest version looks good.

I was going to note that the `displayName()` element on the property is not necessary, but I see that follows the other properties for now, so removing the `displayName()` across the board can be handled separately.

+1 merging

zhtk commented:
I also thank you for reviews and for patience in getting these changes merged 🙇
